### PR TITLE
added Search functionality for subscriptions

### PR DESCRIPTION
**Pull request by tengis617**
**Reference:** #197  

I would like to implement the Search functionality for subscriptions. 
As a starting point, i added the SearchIDs function to subscriptions

### CODE REVIEW

File: testhelpers/slice.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// Contains checks whether a slice of strings contains
+// a given string
+func Contains(list []string, s string) bool {
+	for _, item := range list {
+		if item == s {
+			return true
+		}
+	}
+	return false
+}
```
Comment (leighmcculloch) [inline testhelpers/slice.go]: Could you add a test for this function?

File: subscription_integration_test.go [hunk 2]
```diff
@@ -1129,3 +1129,50 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+	f.Min = time.Now().AddDate(0, 0, -1)
+	result, err := g.SearchIDs(ctx, query)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(result.IDs) < 2 {
+		t.Errorf("expected length of result.IDs to be greater than 2, got %v", len(result.IDs))
+	}
+	if !contains(result.IDs, sub1.Id) {
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: `contains` => `testhelpers.Contains`

File: subscription_integration_test.go [hunk 3]
```diff
@@ -1129,3 +1129,50 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+		t.Errorf("expected result.IDs to include %v", sub1.Id)
+	}
+	if !testhelpers.contains(result.IDs, sub1.Id) {
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: `testhelpers.contains` => `testhelpers.Contains`

File: subscription_integration_test.go [hunk 4]
```diff
@@ -1129,3 +1129,50 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+	g := testGateway.Subscription()
+
+	sub1, err := g.Create(ctx, &SubscriptionRequest{
+		PaymentMethodToken: paymentMethod.GetToken(),
+		PlanId:             "test_plan",
+	})
+	sub2, err := g.Create(ctx, &SubscriptionRequest{
+		PaymentMethodToken: paymentMethod.GetToken(),
+		PlanId:             "test_plan",
+	})
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: Can you add a third subscription that has some property, that will cause it not to be included in the test? To do this you'll need to search on more than `created-at`. This is to ensure that the query is actually being used to filter and we're not just getting back all recent subscriptions.

Comment (tengis617) [reply subscription_integration_test.go]: @leighmcculloch would it be better if i add more tests to check different edge cases? I think just having a single test might actually not be the best idea. Will definitely add the test case you mentioned.

Comment (leighmcculloch) [reply subscription_integration_test.go]: Having a single test with the additional cases within is 👍 .

**Comment by leighmcculloch** [comment]
A test failure was fixed on `master` in #232. Could you merge `master` into your branch?

File: subscription_integration_test.go [hunk 5]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if len(result.IDs) < 2 {
+		t.Errorf("expected length of result.IDs to be greater than 2, got %v", len(result.IDs))
+	}
+	if !testhelpers.Contains(result.IDs, sub1.Id) {
+		t.Errorf("expected result.IDs to include %v", sub1.Id)
+	}
+	if !testhelpers.Contains(result.IDs, sub1.Id) {
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: This line has a typo on it, it is `sub1.Id`, but should be `sub2.Id`.

File: subscription_integration_test.go [hunk 6]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+	f2.Is = "test_plan"
+
+	result, err := g.SearchIDs(ctx, query)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(result.IDs) < 2 {
+		t.Errorf("expected length of result.IDs to be greater than 2, got %v", len(result.IDs))
+	}
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: I think you can delete these three lines. The lines below will error appropriately if only one was returned.

File: subscription_integration_test.go [hunk 7]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+	f2.Is = "test_plan"
+
+	result, err := g.Search(ctx, query)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(result.Subscriptions) < 2 {
+		t.Errorf("expected length of result.IDs to be greater than 2, got %v", len(result.TotalIDs))
+	}
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: You can drop these three lines, per my similar comment above.

File: subscription_integration_test.go [hunk 8]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+		t.Errorf("expected page number to be 1, got %v", result.CurrentPageNumber)
+	}
+	if !testhelpers.Contains(result.TotalIDs, sub1.Id) {
+		t.Errorf("expected subscription ids to contain %v", sub1.Id)
+	}
+	if !testhelpers.Contains(result.TotalIDs, sub2.Id) {
+		t.Errorf("expected subscription ids to contain %v", sub1.Id)
+	}
+	if testhelpers.Contains(result.TotalIDs, sub3.Id) {
+		t.Errorf("expected subscription ids to not contain %v", sub1.Id)
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: Typo, should be: `sub3`.

File: subscription_integration_test.go [hunk 9]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+		t.Errorf("expected length of result.IDs to be greater than 2, got %v", len(result.TotalIDs))
+	}
+	if result.CurrentPageNumber != 1 {
+		t.Errorf("expected page number to be 1, got %v", result.CurrentPageNumber)
+	}
+	if !testhelpers.Contains(result.TotalIDs, sub1.Id) {
+		t.Errorf("expected subscription ids to contain %v", sub1.Id)
+	}
+	if !testhelpers.Contains(result.TotalIDs, sub2.Id) {
+		t.Errorf("expected subscription ids to contain %v", sub1.Id)
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: Typo, should be: `sub2`.

File: subscription_integration_test.go [hunk 10]
```diff
@@ -1129,3 +1129,190 @@ func TestSubscriptionRetryCharge(t *testing.T) {
...
+	result, err := g.Search(ctx, query)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("result.TotalItems = %v", result.TotalItems)
+	t.Logf("result.TotalIDs = %v", result.TotalIDs)
+	t.Logf("result.PageSize = %v", result.PageSize)
+
+	if result.TotalItems < subscriptionCount {
+		t.Fatalf("result.TotalItems = %v, want it to be more than %v", result.TotalItems, subscriptionCount)
```
Comment (leighmcculloch) [inline subscription_integration_test.go]: I think this could be `t.Errorf`. We don't need to fail here do we?

**Comment by tengis617** [comment]
Sorry guys, i was on vacation for a few days. I just fixed all the tests according to the reviews. Hope we can close this soon!

**Comment by leighmcculloch** [comment]
I expect to be able to merge this in Friday. I also plan to cut a new release them.
